core/src/main/java/org/apache/calcite/rel/rules/materialize/MaterializedViewJoinRule.java

```diff
@@ -260,17 +260,20 @@
         : topProject.getProjects();
     List<RexNode> exprsLineage = new ArrayList<>(exprs.size());
     for (RexNode expr : exprs) {
-      Set<RexNode> s = mq.getExpressionLineage(node, expr);
-      if (s == null) {
+      Set<RexNode> lineages = mq.getExpressionLineage(node, expr);
+      if (lineages == null) {
         // Bail out
         return null;
       }
-      assert s.size() == 1;
+      if (lineages.size() != 1) {
+        throw new IllegalStateException("We only support project - filter - join, "
+            + "thus expression lineage should map to a single expression, got: " + lineages);
```

Comment:
Does it make sense to include the input `expr` and or `node` into the message as well?